src/nvapi_gpu.cpp

```diff
@@ -348,4 +348,95 @@ extern "C" {
                 return Error(str::format(n, ": ", adapter->NvmlErrorString(result)));
         }
     }
+
+    NvAPI_Status __cdecl NvAPI_GPU_GetAllClockFrequencies(NvPhysicalGpuHandle hPhysicalGpu, NV_GPU_CLOCK_FREQUENCIES *pClkFreqs) {
+        constexpr auto n = "NvAPI_GPU_GetAllClockFrequencies";
+        static bool alreadyLoggedNoNvml = false;
+        static bool alreadyLoggedHandleInvalidated = false;
+        static bool alreadyLoggedOk = false;
+
+        if (nvapiAdapterRegistry == nullptr)
+            return ApiNotInitialized(n);
+
+        if (pClkFreqs == nullptr)
+            return InvalidArgument(n);
+
+        if (pClkFreqs->version != NV_GPU_CLOCK_FREQUENCIES_VER_1 && pClkFreqs->version != NV_GPU_CLOCK_FREQUENCIES_VER_2 && pClkFreqs->version != NV_GPU_CLOCK_FREQUENCIES_VER_3)
+            return IncompatibleStructVersion(n);
+
+        // Only check for CURRENT_FREQ, and not for the other types ie. BOOST or DEFAULT for now
+        if (pClkFreqs->ClockType != NV_GPU_CLOCK_FREQUENCIES_CURRENT_FREQ)
+            return NotSupported(n);
+
+        auto adapter = reinterpret_cast<NvapiAdapter*>(hPhysicalGpu);
+        if (!nvapiAdapterRegistry->IsAdapter(adapter))
+            return ExpectedPhysicalGpuHandle(n);
+
+        if (!adapter->HasNvml())
+            return NoImplementation(str::format(n, ": NVML not loaded"), alreadyLoggedNoNvml);
+
+        if (!adapter->HasNvmlDevice())
+            return HandleInvalidated(str::format(n, ": NVML available but current adapter is not NVML compatible"), alreadyLoggedHandleInvalidated);
+
+        // Reset all clock data for all domains
+        for (auto i = 4U; i < NVAPI_MAX_GPU_PUBLIC_CLOCKS; i++){
+            pClkFreqs->domain[i].frequency = 0;
+            pClkFreqs->domain[i].bIsPresent = FALSE;
+        }
+
+        unsigned int clock;
+        // Seemingly we need to do nvml call on a "per clock unit" to get the clock
+        // Set the availability of the clock to TRUE and the nvml read clock in the nvapi struct
+        auto resultgpu = adapter->NvmlDeviceGetClockInfo(NVML_CLOCK_GRAPHICS, &clock);
+            switch (resultgpu) {
+                case NVML_SUCCESS:
+                    pClkFreqs->domain[NVAPI_GPU_PUBLIC_CLOCK_GRAPHICS].bIsPresent = TRUE;
+                    pClkFreqs->domain[NVAPI_GPU_PUBLIC_CLOCK_GRAPHICS].frequency = static_cast<long unsigned int>(clock * 1000);
+                    break;
+
+                case NVML_ERROR_NOT_SUPPORTED:
+                    return NotSupported(n);
```

Comment:
I think `Not supported` should just set the domain as not present, thus:
```cpp
pClkFreqs->domain[NVAPI_GPU_PUBLIC_CLOCK_GRAPHICS].bIsPresent = false;
pClkFreqs->domain[NVAPI_GPU_PUBLIC_CLOCK_GRAPHICS].frequency = 0;
```

Reply:
Unless _all_ NVML calls fail with this error, in which case returning `NotSupported` seems to be justified.

Reply:
There is no need to explicitly set bIsPresent to anything, cos i set all to 0/false above with the for loop above - The default state being bIsPresent = false, and frequency = 0. 
Should i just `break:` instead of `return NotSupported(n);`?

Reply:
Yeah, just break seems fine.